esp-wifi/src/wifi/mod.rs

```diff
@@ -1895,209 +1812,155 @@ mod sealed {
-    fn mac_address(self) -> [u8; 6];
-}
-
-/// Wi-Fi station device.
-#[derive(Debug, Clone, Copy, PartialEq)]
-#[cfg_attr(feature = "defmt", derive(defmt::Format))]
-pub struct WifiStaDevice;
 
-impl WifiDeviceMode for WifiStaDevice {
-    fn mode(self) -> WifiMode {
-        WifiMode::Sta
+    fn register_transmit_waker(&self, cx: &mut core::task::Context<'_>) {
+        embassy::TRANSMIT_WAKER.register(cx.waker())
     }
 
-    fn mac_address(self) -> [u8; 6] {
-        let mut mac = [0; 6];
-        sta_mac(&mut mac);
-        mac
+    fn register_receive_waker(&self, cx: &mut core::task::Context<'_>) {
+        match self {
+            WifiDeviceMode::Sta => embassy::STA_RECEIVE_WAKER.register(cx.waker()),
+            WifiDeviceMode::Ap => embassy::AP_RECEIVE_WAKER.register(cx.waker()),
+        }
     }
-}
 
-/// Wi-Fi Access Point (AP) device.
-#[derive(Debug, Clone, Copy, PartialEq)]
-#[cfg_attr(feature = "defmt", derive(defmt::Format))]
-pub struct WifiApDevice;
-
-impl WifiDeviceMode for WifiApDevice {
-    fn mode(self) -> WifiMode {
-        WifiMode::Ap
+    fn register_link_state_waker(&self, cx: &mut core::task::Context<'_>) {
+        match self {
+            WifiDeviceMode::Sta => embassy::STA_LINK_STATE_WAKER.register(cx.waker()),
+            WifiDeviceMode::Ap => embassy::AP_LINK_STATE_WAKER.register(cx.waker()),
+        }
     }
 
-    fn mac_address(self) -> [u8; 6] {
-        let mut mac = [0; 6];
-        ap_mac(&mut mac);
-        mac
+    fn link_state(&self) -> embassy_net_driver::LinkState {
+        match self {
+            WifiDeviceMode::Sta => {
+                if matches!(sta_state(), WifiState::StaConnected) {
+                    embassy_net_driver::LinkState::Up
+                } else {
+                    embassy_net_driver::LinkState::Down
+                }
+            }
+            WifiDeviceMode::Ap => {
+                if matches!(ap_state(), WifiState::ApStarted) {
+                    embassy_net_driver::LinkState::Up
+                } else {
+                    embassy_net_driver::LinkState::Down
+                }
+            }
+        }
     }
 }
 
 /// A wifi device implementing smoltcp's Device trait.
-pub struct WifiDevice<'d, Dm: WifiDeviceMode> {
-    _device: PeripheralRef<'d, crate::hal::peripherals::WIFI>,
-    mode: Dm,
+pub struct WifiDevice<'d> {
+    _phantom: PhantomData<&'d ()>,
+    mode: WifiDeviceMode,
 }
 
-impl<'d, Dm: WifiDeviceMode> WifiDevice<'d, Dm> {
-    pub(crate) fn new(_device: PeripheralRef<'d, crate::hal::peripherals::WIFI>, mode: Dm) -> Self {
-        Self { _device, mode }
-    }
-
+impl WifiDevice<'_> {
     /// Retrieves the MAC address of the Wi-Fi device.
     pub fn mac_address(&self) -> [u8; 6] {
         self.mode.mac_address()
     }
 
+    /// Creates an [smoltcp::iface::Interface] for this [WifiDevice].
+    #[cfg(feature = "smoltcp")]
+    pub fn interface(&mut self) -> smoltcp::iface::Interface {
```

Comment:
When I was reading the examples I thought this was a getter. I propose this should be called `create_interface` instead.

Reply:
If we want any sort of stability in esp-wifi, this would need to be `create_smoltcp_<version>_interface`.